### ISSUE TITLE
sanoid refreshing cache after each dataset when pruning ?

### ISSUE THREAD

**Issue by samarium** [jimsalterjrs/sanoid#202]
seems like sanoid prunes each dataset, then does a force rebuild of the cache?

With in the order of 500 datasets and 40000 snapshots, cache rebuild takes some time, especially where some pools are on HDD.

When I disable $forcecacheupdate, what was talking nearly an hour to complete, finishes in a couple of minutes.

```
diff --git a/sanoid b/sanoid
index d6e58ce..91d7f97 100755
--- a/sanoid
+++ b/sanoid
@@ -241,7 +241,8 @@ sub prune_snapshots {
                                                        }
                                                }
                                                removelock('sanoid_pruning');
-                                               $forcecacheupdate = 1;
+###TMP
+                                               $forcecacheupdate = 0;
                                                %snaps = getsnaps(%config,$cacheTTL,$forcecacheupdate);
                                        } else {
                                                print "INFO: deferring snapshot pruning - valid pruning lock held by other sanoid process.\n";
```
What is going to break if I leave this disabled?
Maybe I need to invoke sanoid --forcecacheupdate when I begin?
Am I going to need to do external locking to stop multiple sanoids running over each other?

**Comment by samarium** [comment]
Further testing, need to comment out the getsnaps call, else cache is still invalidated & reloaded.

With getsnaps 55 minutes, without 3 minutes. Since I run sanoid hourly, 55 minutes is too close.

**Comment by jimsalterjrs** [comment]
Note that if it takes you 52 minutes to generate a list of snapshots, your system is running FAR too lean on RAM and/or you're maintaining far too many snapshots.

Generating a snapshot list is the most common way to bind for hideously long periods of time waiting for metadata, but it's not the only one. Running that lean is likely to bite you sooner or later.

For reference, on a system with enough RAM to handle the number of snapshots you're maintaining, you should be looking at timing more like this:

```
root@banshee:/# time (zfs list -t snap | wc -l)
1903

real	0m0.372s
user	0m0.036s
sys	0m0.328s
```

Sanoid's processing of the list of snapshots does not appreciably add to the time taken:

```
root@banshee:/# time sanoid --force-update --cron

real	0m0.475s
user	0m0.132s
sys	0m0.332s

root@banshee:/# time sanoid --cron

real	0m0.092s
user	0m0.076s
sys	0m0.012s


```
How much RAM do you have on the system you're having trouble with, and how many snapshots are present? Sanoid is pretty smart with its own internal lock process, and will not allow more than one process to fetch snaps and update the cache simultaneously (which keeps a massively overloaded system like yours from death spiraling). But I wouldn't accept any worse than maybe five minutes' time required to get a list of snapshots in production, and I'd be eyeballing ways to cut that few minutes down to more like what you see here, ie a few seconds or less.

Those "ways to cut the time down" pretty much exclusively mean "add more RAM, or keep fewer snapshots".

**Comment by samarium** [comment]
It doesn't take 52 minutes to generate the list of snaps.  It takes about 20-30 seconds to refresh the cache each time.

```
time (zfs list -t snap | wc -l)
35705

real	0m22.863s
user	0m1.196s
sys	0m21.671s
```

System has 32G phys, 8G ARC, 10G free.  When I increase ARC max, it doesn't grow quickly, even when I'm running sanoid --cron.  CPU load is generally < 20%, so neither RAM or CPU seem to be limiting factors.  Pools are a combination of SSD mirror, HDD mirror, and HDD Z2 with SSD ZIL & L2ARC, not that the L2ARC is doing much, but ZIL is useful for NFS.

The problem seems to be that each time a set of snapshots for a dataset is pruned, the cache is updated again, rather than just using the existing cache, just like the code above from sanoid above, even when I set $forcecacheupdate to 0.  From my so far initial look at the code I would expect the cache to be valid for $cacheTTL=900 otherwise, but that doesn't seem to be the case.  But I expect you set forcecacheupdate=1 on purpose to have a more coherent cache, however I don't understand the implications yet.

If you have 1903 snapshots, with approx 50 snaps/dataset if you use something like your defaults, then you have about 40 datasets.

While I have about 200 datasets that I'm running thru sanoid in this run, there are another 250 or so on the system, but unfortunately when sanoid refreshes it's cache it is looking at all the snapshots on the system, not just the ones it is interested in.

Still, 35k snaps seems large.  Given 500 datasets, 50 snaps/dataset, is 25k, so right order of magnitude,  and I'm in the middle of transitioning between one snap system to another, so that explains the snap number.

Looking at stock sanoid --cron --verbose output where it transitions from taking to pruning I see this:
```
taking snapshot tank/vm/vmosx1@autosnap_2018-02-22_13:03:25_hourly
INFO: cache expired - updating from zfs list.
INFO: pruning snapshots...
INFO: pruning tank/vm/pool@autosnap_2018-02-20_23:02:03_hourly ... 
INFO: cache expired - updating from zfs list.
INFO: pruning tank/software@autosnap_2018-02-20_23:02:03_hourly ... 
INFO: cache expired - updating from zfs list.
```
and the cache is refreshed before pruning each dataset.

I currently have 4 pools active, and looking at the list snap duration on each:
```
real 0m00.304s  SSD mirror
real 0m04.444s  HDD Z2 + SSD ZIL / L2ARC, 7200rpm
real 0m11.545s  HDD mirror, 5400rpm
real 0m08.261s  HDD mirror, 7200rpm
```

Even with just the first 2, which are what I'm running thru sanoid --cron, I don't think I really want the cache reloading after every dataset prune.

**Comment by jimsalterjrs** [comment]
There are probably some answers there for more intelligently deciding 
when to or not to update the cache, but you have to be very, very 
careful with it.

The problem is that right now, Sanoid never alters existing cache, it 
merely regenerates it from scratch. Once you realize that, you also 
realize that Sanoid can't actually know how many snapshots of a 
particular type are present for a given dataset (and thus, whether it 
should remove more of them) unless it updates its cache, which, like I 
said, currently requires actually regenerating it. If you assume you 
have every snapshot in cache, you run the risk of pruning a snapshot you 
shouldn't have, because you think you've got 50 and you've actually only 
got /n/ due to previous pruning.

It gets even crazier when you realize that you /may/ have multiple 
Sanoid processes running - and that's a feature, not a bug; they have 
inter-process locking to avoid trying to update the cache or take 
snapshots at the same time, but you don't want a new snapshot to /not/ 
get taken on time because another Sanoid process is running long trying 
to regenerate the cache, etc.

You're encountering something of a corner case right now; I don't think 
most users have anywhere near as many datasets as you do, so won't 
encounter the 250x re-generating the cache in one run behavior that you are.

**Comment by phreaker0** [comment]
I'm also affected by this, sanoid will take very long to run. In my case the server is highly loaded (vm host) but disk io is pretty low and there is enough free memory, but snapshot listing like sanoid does it takes almost a minute. Running it several times after another doesn't speed things up (as you can see in the following log I have a huge number of snapshots ~50000, with about ~230 datasets):

```
root@sam:~# date; time zfs get -Hpt snapshot creation | wc
Thu Feb 22 08:41:27 CET 2018
  52332  209328 5465966

real    0m52.518s
user    0m2.372s
sys     0m40.340s
root@sam:~# date; time zfs get -Hpt snapshot creation | wc
Thu Feb 22 08:42:19 CET 2018
  52331  209324 5465892

real    0m51.952s
user    0m2.264s
sys     0m40.672s
root@sam:~# date; time zfs get -Hpt snapshot creation | wc
Thu Feb 22 08:43:11 CET 2018

  52330  209320 5465822

real    0m50.124s
user    0m2.404s
sys     0m40.580s

root@sam:~# arcstat 
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  arcsz     c  
08:44:17   843   817     96   816   96     0    0   816   96   3.4G   12G
root@sam:~# free -m
              total        used        free      shared  buff/cache   available
Mem:          32088       17025       12136         355        2925       14249
Swap:             0           0           0
```

**Comment by samarium** [comment]
@phreaker0 good to see I'm no the only phreak with lots of datasets :-)

My current work around is to comment the getsnaps() in the code fragement above, and ensure I'm not running multiple sanoid prunes on the same datasets.  New snapshots I'm not so worried about, as it would just mean I would prune less than optimum.

**Comment by phreaker0** [comment]
skipping cache regeneration after each prune has an huge impact for me, sanoid run times are down from 20-55 minutes to 1-2 minutes:

```
Feb 22 07:21:17 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 08:14:01 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 08:14:01 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 08:58:12 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 08:58:12 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 09:39:31 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 09:39:31 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 09:58:01 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 09:58:01 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 10:41:09 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 10:41:09 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 11:13:39 sam systemd[1]: Started Snapshot ZFS Pool.

# patched sanoid (don't regenerate cache file after pruning)

Feb 22 11:13:39 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 11:15:37 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 11:15:37 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 11:17:09 sam systemd[1]: Started Snapshot ZFS Pool.

Feb 22 11:30:00 sam systemd[1]: Starting Snapshot ZFS Pool...
Feb 22 11:31:20 sam systemd[1]: Started Snapshot ZFS Pool.

```

So this should be properly handled. What about removing the destroyed snapshots from the cachefile instead of regenerating them @jimsalterjrs?

@samarium if one runs sanoid in parallel and one of the instances generates the cachefile, the other one will skip the generation and use the old one anyhow (which probably contains snapshots which no longer exists)

**Comment by jimsalterjrs** [comment]
In the simplest band aid sense, we could probably just set a 
datasets_pruned flag that makes sure we update cache before exit, rather 
than naively updating it after each individual dataset is pruned the way we 
do now.

What would be better is improving the way we update and share the cache 
between concurrently running processes. This might involve keeping a 
separate hash of snapshots pruned and snapshots taken by the running 
process, as well as editing the cache in memory. Doing it this way, we'd be 
free to rewrite the cache on disk each time we updated it, just making sure 
before writing it that we checked to see if it was newer than our current 
version in memory.

If we found a newer cache on disk than the one we have in memory (not 
counting our own live edits), we could load the fresher cache into the 
running process, apply the changes we'd already made (snapshots taken or 
deleted) to that imported cache, then write our updated version back out.

This would allow any sanoid process the benefit of the latest information 
while requiring regeneration of cache far less frequently.

Thoughts? Am I making sense here?

**Comment by phreaker0** [comment]
For the band aid fix I wouldn't update the cache on exit but just invalidate it by deleting the cache file altogether. If the same process needs it again it will automatically created again as will the next one.

I think the way i would update the cache file after pruning will match your description:

- check mutex 
- get mutex
- read cache file to memory
- remove the destroyed snapshots
- write cache file back to filesystem
- release lock

so the only difference to the current way is that instead of running the zfs command, the cache file will be update in memory and written back. This still would be done after each dataset like it is now, but will be much faster and would provide other sanoid instances faster with an updated cache file.

**Comment by jimsalterjrs** [comment]
> I think the way i would update the cache file after pruning will match 
> your description:
>
>   * check mutex
>   * get mutex
>   * read cache file to memory
>   * remove the destroyed snapshots
>   * write cache file back to filesystem
>   * release lock
>
> so the only difference to the current way is that instead of running 
> the zfs command, the cache file will be update in memory and written 
> back. This still would be done after each dataset like it is now, but 
> will be much faster and would provide other sanoid instances faster 
> with an updated cache file.
>
What happens when one Sanoid process is pruning snapshots, and another 
fires up to take them? You're going to get conflicting incorrect 
versions of the cache. One will be missing the new snapshot (causing 
multiples to be taken), one will incorrectly still contain the 
already-pruned snapshots (causing failed attempts to remove snapshots 
that are already gone - this already happens sometimes).

Which one of those incomplete cache versions "sticks" just depends on 
the last process to overwrite it.

To handle this /correctly/, each process really needs to be able to 
track its own modifications, so that it can re-read cache from disk when 
necessary and then patch it immediately with the changes it's already 
made, as necessary.

**Comment by samarium** [comment]
If saniodT is taking snapshots, then it will do so if the time is right, and pruning won't matter that much if there are at least a few snaps in the list.

If snapoidP is pruning snapshops, then it might leave one that could have been pruned that as sanoidT was adding snapshots to the other end of the list, again, not that much of a problem, syncoid has probably already synced that left over one already anyway.

If syncoid is running, then maybe there are a few snaps getting replicated a little slower, but can't see it being much of a problem.

I am running different sanoid instances from different configdirs depending on which pools are connected, so I have relatively fine control and now all sanoids easily complete in the top of the hour.  I have day snapshot time as 00:00, so hourly is fine, and hourly interval is fine too, don't need to run every minute or 5.

I'm only running syncoid 3x /day, and if I miss one run then can complete the sync next time.  If that became a problem, then 4x or 6x would be fine too.

**Comment by phreaker0** [comment]
> What happens when one Sanoid process is pruning snapshots, and another fires up to take them? You're going to get conflicting incorrect versions of the cache. One will be missing the new snapshot (causing multiples to be taken), one will incorrectly still contain the already-pruned snapshots (causing failed attempts to remove snapshots that are already gone - this already happens sometimes). Which one of those incomplete cache versions "sticks" just depends on the last process to overwrite it. To handle this /correctly/, each process really needs to be able to track its own modifications, so that it can re-read cache from disk when necessary and then patch it immediately with the changes it's already made, as necessary.

Exactly the same conflicts that can happen now, will happen again but there shouldn't be more. If the cache update lock can't be obtained after a pruning iteration it will be deferred like before and all the removed snapshots are queued for later cache removal. If all prune iterations are done and there are still cached snapshots which need to be removed, sanoid will wait for the lock as long as it takes and removes those entries finally (if they are still there, could be that another process already made a new zfs get call ...)
I will create a PR, maybe the code is clearer than my explanations :-D

**Comment by samarium** [comment]
Thanks @phreaker0 I grabbed your patches and am running that sanoid.  First run was OK.

Not sure that I'm a good test of cache coherency, however sanoid finished 4 minutes after starting, which is much better than 50+ minutes, so largely the same runtime as my quick hack, certainly within my 10 minute target.

As for concurrency, I do potentially have 2 sanoids running, but they prune different pools, and are offset 3 minutes, and the first only runs 6hourly  so not much overlap, although I expect occassional heavy disk IO would cause more concurrency.

I'll see how it goes over the next week.
